### PR TITLE
Removing objdump and using light_elf.h

### PR DESCRIPTION
**Pull request by antony-jr**
This **PR** replaces the system specific code for retriving **type 2 AppImage** update information using **objdump** with much better solution by using **light_elf.h**. 
This also avoids **bundling objdump** in the AppImage.

### CODE REVIEW

File: resources/build-appimages.sh [hunk 1]
```diff
@@ -41,8 +41,7 @@ cp -v "$REPO_ROOT"/resources/*.desktop AppDir/usr/share/applications/
 cp -v "$REPO_ROOT"/resources/*.svg AppDir/usr/share/icons/hicolor/scalable/apps/
 cp -v "$REPO_ROOT"/resources/*.xpm AppDir/resources/
 
-# bundle objdump required for parsing update information
-cp /usr/bin/objdump AppDir/usr/bin/
+# no need for objdump to be bundled!
```
Comment (TheAssassin) [inline resources/build-appimages.sh]: Exclamation marks are not necessary for comments.

File: src/updater.cpp [hunk 2]
```diff
@@ -139,48 +139,44 @@ namespace appimage {
...
-                        auto pathToBinary = std::string(buffer.data());
-                        auto slashPos = pathToBinary.find_last_of('/');
-                        auto bundledObjdump = pathToBinary.substr(0, slashPos) + "/objdump";
-
-                        if (isFile(bundledObjdump)) {
-                            objdump = bundledObjdump;
-                        } else {
-                            objdump = "objdump";
+                   {
+                     unsigned offset = 0, length = 0;
```
Comment (TheAssassin) [inline src/updater.cpp]: This code should be moved to a separate function so it can be reused.

File: src/updater.cpp [hunk 3]
```diff
@@ -139,48 +139,44 @@ namespace appimage {
...
+                     uint8_t * data;
```
Comment (TheAssassin) [inline src/updater.cpp]: Breaks with code style. Pointers are instantiated like `uint8_t* data;`.

File: src/updater.cpp [hunk 4]
```diff
@@ -139,48 +139,44 @@ namespace appimage {
...
+                     Elf64_Ehdr * elf64;
+                     Elf64_Shdr * shdr64;
+                     Elf32_Ehdr * elf32;
+                     Elf32_Shdr * shdr32;
+                     int i, fd = open(pathToAppImage.c_str(), O_RDONLY);
+                     data = (uint8_t * ) mmap(NULL, lseek(fd, 0, SEEK_END), PROT_READ, MAP_SHARED, fd, 0);
```
Comment (TheAssassin) [inline src/updater.cpp]: This should be replaced by a less error prone C++ solution.
Bracket use breaks code style.

File: src/updater.cpp [hunk 5]
```diff
@@ -139,48 +139,44 @@ namespace appimage {
...
-                        } else {
-                            objdump = "objdump";
+                   {
+                     unsigned offset = 0, length = 0;
+                     uint8_t * data;
+                     Elf64_Ehdr * elf64;
+                     Elf64_Shdr * shdr64;
+                     Elf32_Ehdr * elf32;
+                     Elf32_Shdr * shdr32;
+                     int i, fd = open(pathToAppImage.c_str(), O_RDONLY);
```
Comment (TheAssassin) [inline src/updater.cpp]: Empty lines are not a crime. Perhaps you should insert some.

File: src/updater.cpp [hunk 6]
```diff
@@ -139,48 +139,44 @@ namespace appimage {
...
+                     data = (uint8_t * ) mmap(NULL, lseek(fd, 0, SEEK_END), PROT_READ, MAP_SHARED, fd, 0);
+                     elf64 = (Elf64_Ehdr * ) data;
+                     shdr64 = (Elf64_Shdr * )(data + elf64-> e_shoff);
+                     if (elf64-> e_ident[EI_CLASS] != ELFCLASS64) {
```
Comment (TheAssassin) [inline src/updater.cpp]: Missing indentation.

File: src/updater.cpp [hunk 7]
```diff
@@ -139,48 +139,44 @@ namespace appimage {
...
+                     elf32 = (Elf32_Ehdr * ) data;
+                     shdr32 = (Elf32_Shdr * )(data + elf32-> e_shoff);
```
Comment (TheAssassin) [inline src/updater.cpp]: Use of `->` operator breaks with code style.

File: src/updater.cpp [hunk 8]
```diff
@@ -139,48 +139,44 @@ namespace appimage {
...
+
+                     char * strTab = (char * )(data + shdr32[elf32-> e_shstrndx].sh_offset);
+                     for (i = 0; i < elf32-> e_shnum; i++) {
+                     if (strcmp( & strTab[shdr32[i].sh_name], ".upd_info") == 0) {
+                        offset = shdr32[i].sh_offset;
+                        length = shdr32[i].sh_size;
+                      }
+                     }
+                     } else {
```
Comment (TheAssassin) [inline src/updater.cpp]: Follow-up for comment in line 153: if the block above this line would be properly indented, this was much easier to read.

File: src/updater.cpp [hunk 9]
```diff
@@ -139,48 +139,44 @@ namespace appimage {
...
+                       char * strTab = (char * )(data + shdr64[elf64-> e_shstrndx].sh_offset);
+                       for (i = 0; i < elf64-> e_shnum; i++) {
+                         if (strcmp( & strTab[shdr64[i].sh_name], ".upd_info") == 0) {
```
Comment (TheAssassin) [inline src/updater.cpp]: Use of `&` operator breaks code style.

File: src/updater.cpp [hunk 10]
```diff
@@ -139,48 +139,44 @@ namespace appimage {
...
+                             offset = shdr64[i].sh_offset;
+                             length = shdr64[i].sh_size;
+                          }
                         }
+                      }
+                     close(fd);
+                     char update_info[length];
```
Comment (TheAssassin) [inline src/updater.cpp]: Variable name breaks with code style.

File: src/updater.cpp [hunk 11]
```diff
@@ -139,48 +139,44 @@ namespace appimage {
...
-                    auto offset = (unsigned long) std::stoi(parts[5], nullptr, 16);
-                    auto length = (unsigned long) std::stoi(parts[2], nullptr, 16);
-
-                    ifs.seekg(offset, std::ios::beg);
-                    std::vector<char> rawUpdateInformation(length, '\0');
-                    ifs.read(rawUpdateInformation.data(), length);
-
-                    updateInformation = rawUpdateInformation.data();
-                } else {
+		} else {
```
Comment (TheAssassin) [inline src/updater.cpp]: Indentation changed, now broken (not a multiple of 4).

File: src/util.h [hunk 12]
```diff
@@ -7,13 +7,25 @@
 #include <sstream>
 #include <string>
 #include <vector>
+#include <stdio.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include "light_elf.h"
```
Comment (TheAssassin) [inline src/util.h]: Breaks with code style for includes. They're not sorted alphabetically, and `light_elf.h` is a local include, not a system one.

File: src/util.h [hunk 13]
```diff
@@ -7,13 +7,25 @@
...
 
 namespace appimage {
     namespace update {
-        static void removeNewlineCharacters(std::string& str) {
-            str.erase(std::remove(str.begin(), str.end(), '\n'), str.end());
-        }
+        static std::vector<std::string> split(const std::string& s, char delim = ' ') {
```
Comment (TheAssassin) [inline src/util.h]: You moved this function up for some reason I don't understand. Please move it to its original position.

File: src/util.h [hunk 14]
```diff
@@ -22,7 +34,7 @@ namespace appimage {
             }));
             return s.length() < initialLength;
         }
-
+
```
Comment (TheAssassin) [inline src/util.h]: Spaces on empty lines should be removed.

File: src/util.h [hunk 15]
```diff
@@ -104,4 +72,4 @@ namespace appimage {
             return (bool) ifs && ifs.good();
         }
     }
-}
```
Comment (TheAssassin) [inline src/util.h]: Good find, my IDE sometime refuses to notify me of missing trailing newline characters.

**Comment by TheAssassin** [comment]
I mention some code style issues once only, but they can be found all over the place. Please fix those everywhere you see them.

Also, it looks like you "stole" the code in AppImageKit's code base. If that's the case, this breaks AppImageKit's license. You have to fix that asap.

**Comment by TheAssassin** [comment]
That said, thanks a lot for your PR. This has been long overdue, and I had it on the TODO list for the next sprint since last month.

**Comment by antony-jr** [comment]
Can I use **gnu/indent** for indentation and to reformat other contents with your code style? if so . Can you please give some code style reference.
Example:= To use linux coding style ``` indent -linux source.cpp ```

**Comment by TheAssassin** [comment]
I am not using any such auto code format tools. But the files are formatted pretty consistently, you can just have a look at the rest of the code to get an idea. Also, I've listed every violation of the code style I found, that makes it easier to fix them.

**Comment by probonopd** [comment]
In laymans terms, what is wrong with objdump so that you think it should be removed? It is working. Changing it can only introduce more issues than we have currently (=none), right? I am always hesitant to make such changes to working systems when there is no real pressing need.

**Comment by antony-jr** [comment]
> what is wrong with objdump so that you think it should be removed?

well this can be considered as good practice. This also makes the complexity less. Like , When you use **objdump** , you have to package it with your appimage and then do **text manipulation** with the output of **subprocess**. From my experience , Using subprocess , in other words , Using external program should be our last resort.

And If there is any case of static linking to a single executable without AppImage support how could we handle this ? Prehaps in case of **AppImageSelfUpdate**.

>  It is working. Changing it can only introduce more issues than we have currently with it (=none), right?

But we cannot go on with **subprocess** and this merge will not cause any issues later (After I make the **changes requested** ). Infact it will reduce some future issues , like security , license and performance.

**Comment by TheAssassin** [comment]
> In laymans terms, what is wrong with objdump so that you think it should be removed? It is working. Changing it can only introduce more issues than we have currently with it (=none), right? I am always hesitant to make such changes to working systems when there is no real pressing need.

The main problem is that in some cases, the objdump output parser tended to be buggy, causing AppImageUpdate to get stuck.

> well this can be considered as good practice. This also makes the complexity less.

I think that's not really an issue here.

> Like , When you use objdump , you have to package it with your appimage

This is a real problem for applications building upon libappimageupdate. They don't know they have to bundle objdump in most cases I'd say. Replacing it with a standalone solution seems like a better solution.

> and then do text manipulation with the output of subprocess.

Well, "text manipulation" is probably the wrong term. You're just parsing data out of a process's output. That's the every day business of software developers.

> From my experience , Using subprocess , in other words , Using external program should be our last resort.

Agree. That's why there was a TODO in the code.

> If there is any case of static linking to a single executable without AppImage support how could we handle this ? Prehaps in case of AppImageSelfUpdate.

Not sure what this is supposed to tell us.

>> It is working. Changing it can only introduce more issues than we have currently with it (=none), right?
> 
>  But we cannot go on with subprocess and this merge will not cause any issues later (After I make the changes requested ). Infact it will reduce some future issues , like security , license and performance.

There's absolutely no licensing issues here. GPL software can be redistributed in binary form freely. There's also no security issues as far as I can see, neither are there performance issues (it's a one time call to a subprocess, and such calls are blazing fast on unixes -- bash scripts work like this, too, and they can be really fast as well).

**Comment by antony-jr** [comment]
Will reopen after making the requested changes , I made a lot of conflicts and I think I would do the changes in a fresh copy and then give a PR. ❤️

File: src/updater.cpp (hunk 4 above)
Comment (antony-jr) [reply src/updater.cpp]: What Can I replace mmap in C++11 ? 

In Qt I can simply do this to map a file safely...
```
QFile AppImage("Some.AppImage");
if(AppImage.open(QIODevice::ReadOnly)){
  uchar *Mapped = AppImage.map(0 , AppImage.size());
  uint8_t *data = (uint8_t*) Mapped;
  AppImage.unmap(Mapped);
  AppImage.close();
}
```

Any Idea's ?

Comment (TheAssassin) [reply src/updater.cpp]: Sorry for not having been clear about this.

I don't like `mmap()` in general, as mapped files could, in theory, be altered in the background, hence changing the data in the map as well, which can cause some hard to debug bugs. Reading files using `std::ifstream` or even the C API (`read()` etc.) doesn't have such problems.

But as we're a) just reading from the file and b) the ELF API expects a char pointer, it is okay as-is... (I wouldn't have built it that way, but well...). So, sorry for being unclear, there is actually no issue here, I didn't read the rest of the code below.

Looking at the rest of the code, this seems safe to use. I just have a question: is the file unmapped at some point? I don't see a `munmap` call anywhere.

P.S.: My pointer-free alternative to using `mmap()` or `read()` or anything alike:

```cpp
#include <algorithm>
#include <fstream>
#include <string>

int main() {
    // ifstream doesn't provide any write facilities
    std::ifstream ifs("testfile");

    // checks if file could be opened
    if (ifs) {
        // initialize a new vector and resize it to the file size
        auto fileSize = ifs.tellg();
        std::vector<char> fileData(fileSize);

        // this is how you could seek to file's beginning
        // however, since we just opened the file, we're reading from the beginning already
        //ifs.clear();
        //ifs.seekg(0, std::ios::beg);

        // now, read in data
        ifs.read(fileData.data(), fileSize);

        // do something meaningful
        // ...

        return 0;
    }

    return 1;
}
```

Since data is a stack object, it will be deleted automatically once its scope ends. No cleanup required at all, no memory leaks, etc.. Pretty convenient to use, in my opinion.

Comment (antony-jr) [reply src/updater.cpp]: > Looking at the rest of the code, this seems safe to use. I just have a question: is the file unmapped at some point? I don't see a ```munmap``` call anywhere.

Yes , I forgot to unmap the mapped! 

Then I will be using **fstream** to find the filesize instead of the **C API**. And will use stack memory instead of heap for the **read** calls from **fstream**.

One thing that bothers me is that we have to open the file twice ( One for the stream and another for the **EFI API** ). (In Qt there is only one call to the file ).
For this , I found some compiler specific solutions on fstream using a file descriptor ( Can't use it because its gcc specific)

Thank You for your quick response! 👍

Comment (TheAssassin) [reply src/updater.cpp]: > use stack memory

For the sake of completeness: I am pretty sure the `std::vector<...>` class does allocate its underlying arrays on the heap. 

Just keep using the C API in this case, if you like. Introducing an `fstream` just because we can doesn't make sense. Especially since we have a C file pointer anyway. As said before, it was my mistake, as I didn't fully understand what you're doing here at first.

> EFI

ELF *
